bamboolean/lexer.py

```diff
@@ -105,65 +107,51 @@ def _integer(self):
             self.next()
         return int(result)
 
+    def skip_and_invoke(self, count, func):
+        for i in range(count):
+            self.next()
+        return func()
+
+    def is_token_equal(self, expected):
+        return expected == reduce(
+            lambda actual, _: actual + str(self.peek()),
+            [self.current_char] * len(expected),
+        )
+
     def get_next_token(self):
         """
-        Lexical anaylzer (tokenizer). Breaks sentence apart into tokens
+        Lexical analyzer (tokenizer). Breaks sentence apart into tokens
         """
         while self.current_char is not None:
             if self.current_char.isspace():
                 self.skip_whitespace()
                 continue
 
-            if re.match('[_a-zA-Z]', self.current_char):
-                return self.id()
-
-            if self._is_quotation_mark(self.current_char):
-                self.next()
-                return self.string()
-
-            if self.current_char.isdigit():
-                return self.number()
-
-            # Relational operators
-
-            if self.current_char is '=' and self.peek() == '=':
-                self.next()
-                self.next()
-                return Token(tok.EQ, '==')
-
-            if self.current_char is '!' and self.peek() == '=':
-                self.next()
-                self.next()
-                return Token(tok.NE, '!=')
-
-            if self.current_char is '<':
-                # LTE, LT
-                self.next()
-                if self.current_char is '=':
-                    self.next()
-                    return Token(tok.LTE, '<=')
-                else:
-                    return Token(tok.LT, '<')
-
-            if self.current_char is '>':
-                # GTE, GT
-                self.next()
-                if self.current_char is '=':
-                    self.next()
-                    return Token(tok.GTE, '>=')
-                else:
-                    return Token(tok.GT, '>')
-
-            # Parentheses
-
-            if self.current_char == '(':
-                self.next()
-                return Token(tok.LPAREN, '(')
-
-            if self.current_char == ')':
-                self.next()
-                return Token(tok.RPAREN, ')')
-
+            regex_map = OrderedDict((
```

Comment:
The way it works is slightly odd - you have a regexp . When I saw regexp my expectation was that it would interpret the whole string/id/number (like in this implementation: https://gist.github.com/eliben/5797351)

It may be more obvious if you do something like that or alternatively instead of regexp you just had a dict from the first character to the function (after all the dispatch based on the first character is what happens here).

Reply:
I see, in my implementation, it's just about recognizing what am i dealing with (number? string? etc). Dict from the first character to the function - you mean listing all initial characters? but that would make it the size of the alphabet, or did i misunderstood you?

The lexer that you attached is neat indeed, we may refactor our for using similar pattern. However, i have to say that it is outside of the scope of this commit, since it would pretty much involve rewriting whole lexer.

Reply:
> Dict from the first character to the function - you mean listing all initial characters?
Not explicitly.
More like:
```
dispatch_dict = dict(
  [(c, self.id) for c in self.identifier_chars] +
  [(c, self.string) for c in ["'", '"']] +
  [(c, self.number) in self.digits)
)
```

Not sure if worth it.

Reply:
I would still need to define `self.identifier_chars = "_abcde..."`.  Also i don't think that improves the readability well. It also does not scale if we would like to make prediction based on first two characters - now it may be achieved similarly as function for `tokens_map` is done (with `peek()` function)

Reply:
Well, actually if you had a string instead of a single character it would work just as well (there would need to be a loop over.  I don't see any problem with that. Assuming the sentinels are set up properly it's as simple as `self.text[self.current_pos:len(s)] == s`. Actually you could unify it with `token_map` and that's a good thing... (And if performance is ever and issue and it's not rewritten 17 times by then, you can make it a TRIE instead of iterating over all of them).